extensions/amp-social-share/0.2/amp-social-share.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {Layout} from '../../../src/layout';
+import {PreactBaseElement} from '../../../src/preact/base-element';
+import {Services} from '../../../src/services';
+import {SocialShare} from './social-share';
+import {addParamsToUrl} from '../../../src/url';
+import {dict} from '../../../src/utils/object';
+import {getDataParamsFromAttributes} from '../../../src/dom';
+import {getSocialConfig} from './amp-social-share-config';
+import {isExperimentOn} from '../../../src/experiments';
+import {toggle} from '../../../src/style';
+import {user, userAssert} from '../../../src/log';
+
+/** @const {string} */
+const TAG = 'amp-social-share';
+
+class AmpSocialShare extends PreactBaseElement {
+  /** @override */
+  init() {
+    const viewer = Services.viewerForDoc(this.element);
+    const platform = Services.platformFor(window);
+    const type = userAssert(
```

Comment:
We should see if we can immediately allow (a) creating this element without `type`, and (b) changing this type later.

Reply:
Neither of these are part of the API now, is there a particular reason you want to support these?

Reply:
Can you clarify this? We currently have `<amp-social-share type="...">`, right? The goal is to make everything mutable. For instance, we've already been fielding the complaints that it's impossible to create some elements, e.g. `document.createElement('amp-social-share')` doesn't work since it's impossible to set attributes in `createElement` API.

Reply:
`document.createElement('amp-social-share')` itself appears to be OK, only errors requiring `type` if you attach this to the document before setting the attribute.

This works:
```
const mySocialShare = document.createElement('amp-social-share');
mySocialShare.setAttribute('type', 'facebook');
document.documentElement.appendChild(mySocialShare);
```

This throws an error and results in an unbuilt element:
```js
const mySocialShare = document.createElement('amp-social-share');
document.documentElement.appendChild(mySocialShare);
setTimeout(() => {
    mySocialShare.setAttribute('type', 'facebook');
}, 500);
```

Reply:
Ok, this part is ok then. However, do we have any value otherwise in restricting this?

Reply:
AFAICT because [non-configured providers are allowed](https://amp.dev/documentation/components/amp-social-share/#non-configured-providers), this is essentially equivalent to allowing no `type` attribute because the given value can also be meaningless and the component will still work.

Reply:
I think this is OK as is. If you were to create the component without a type, we should be throwing. We could consider moving that into the Comp, but we don't have a good assertion story for Preact yet.

Reply:
I'm seeing mostly that if a React Component doesn't like something, it simply does `return null` and maybe prints a warning. Asserts seem rather rare.